In [1]:
import pandas as pd
import numpy as np
import math

In [2]:
data = pd.read_csv('Islander_data.csv')

In [3]:
happy = data[data['Happy_Sad_group'] == 'H']

In [4]:
happy_drug_a = happy[happy['Drug'] == 'A']
happy_drug_a_dosage_2 = happy_drug_a[happy_drug_a['Dosage'] == 2]
happy_drug_a_dosage_2

first_name  last_name  age Happy_Sad_group  Dosage Drug  Mem_Score_Before  \
23       Anna    Bernard   62               H       2    A              46.5   
25       Dean   Carrasco   40               H       2    A              47.0   
28      Sofia   Carrasco   28               H       2    A              43.9   
29     Darren     Durand   36               H       2    A              74.9   
31     Misaki     Durand   29               H       2    A              58.9   
33     Robert     Durand   35               H       2    A              58.8   
35      Ayano    Kennedy   38               H       2    A              40.2   
37       Jose      Lopez   25               H       2    A              50.0   
40    Anthony    Steiner   26               H       2    A              87.0   
41       Nina    Steiner   51               H       2    A              46.5   
44     Felipe  Takahashi   28               H       2    A              60.9   

    Mem_Score_After  Diff  
23             46.1  -0.4  
25             55.5   8.5  
28             46.8   2.9  
29             70.8  -4.1  
31             56.3  -2.6  
33             50.8  -8.0  
35             44.5   4.3  
37             49.4  -0.6  
40            102.0  15.0  
41             50.8   4.3  
44             64.9   4.0

In [5]:
happy_control = happy[happy['Drug'] == 'S']
happy_control_dosage_2 = happy_control[happy_control['Dosage'] == 2]
happy_control_dosage_2

first_name  last_name  age Happy_Sad_group  Dosage Drug  Mem_Score_Before  \
89   Christian   Carrasco   28               H       2    S              36.0   
90     Ignacio   Carrasco   46               H       2    S              54.8   
93        Sara     Castro   35               H       2    S              46.0   
94    Benjamin   Connolly   27               H       2    S              67.0   
98         Rin     Durand   40               H       2    S              30.7   
99         Rok      Hajek   48               H       2    S              48.3   
102    Bastian      Lopez   42               H       2    S              46.2   
103      Kaito      Lopez   37               H       2    S              41.0   
105     Victor      Price   28               H       2    S              79.7   
107       Kana    Summers   51               H       2    S              85.5   
110       Shun  Takahashi   38               H       2    S              69.2   

     Mem_Score_After  Diff  
89              35.8  -0.2  
90              65.4  10.6  
93              43.2  -2.8  
94              70.9   3.9  
98              27.1  -3.6  
99              47.0  -1.3  
102             41.3  -4.9  
103             47.0   6.0  
105             88.9   9.2  
107             83.6  -1.9  
110             65.8  -3.4

In [6]:
control, experimental = happy_control_dosage_2['Diff'], happy_drug_a_dosage_2['Diff']
df = pd.DataFrame([control.tolist(), experimental.tolist()])
#df.transpose().to_csv('data.csv', header=['control', 'experimental'], index=False)
df

0     1    2    3    4    5    6    7     8    9   10
0 -0.2  10.6 -2.8  3.9 -3.6 -1.3 -4.9  6.0   9.2 -1.9 -3.4
1 -0.4   8.5  2.9 -4.1 -2.6 -8.0  4.3 -0.6  15.0  4.3  4.0

In [7]:
import matplotlib.pyplot as plt

fig, ax = plt.subplots()


x1, y1 = np.arange(1, 12) - 0.2, control
x2, y2 = np.arange(1, 12) + 0.2, experimental

ax.bar(x1, y1, width = 0.4)
ax.bar(x2, y2, width = 0.4)
ax.set_facecolor('seashell')

fig.set_figwidth(12)    
fig.set_figheight(6)
fig.set_facecolor('floralwhite')

plt.show()

<Figure size 1200x600 with 1 Axes>

Гипотеза H0: θ = 0

Гипотеза H1: θ > 0

In [8]:
ranges = list(zip(control, ['c'] * 11)) + list(zip(experimental, ['e'] * 11))
ranges = list(zip(enumerate(sorted(ranges, key = lambda x: abs(x[0])))))
ranges_df = pd.DataFrame([
    list(range(1,23)),
    [x[0][1][0] for x in ranges],
    [x[0][1][1] for x in ranges],
])
ranges_df

0    1    2    3    4    5    6    7    8    9   ...   12   13   14   15  \
0    1    2    3    4    5    6    7    8    9   10  ...   13   14   15   16   
1 -0.2 -0.4 -0.6 -1.3 -1.9 -2.6 -2.8  2.9 -3.4 -3.6  ... -4.1  4.3  4.3 -4.9   
2    c    e    e    c    c    e    c    e    c    c  ...    e    e    e    c   

   16  17   18   19    20  21  
0  17  18   19   20    21  22  
1   6  -8  8.5  9.2  10.6  15  
2   c   e    e    c     c   e  

[3 rows x 22 columns]

In [9]:
exp_ranges_neg = [x[0][0] + 1 for x in ranges if x[0][1][1] == 'e' and x[0][1][0] < 0]
exp_ranges_neg, sum(exp_ranges_neg)

([2, 3, 6, 13, 18], 42)

In [10]:
exp_ranges_pos = [x[0][0] + 1 for x in ranges if x[0][1][1] == 'e' and x[0][1][0] > 0]
exp_ranges_pos, sum(exp_ranges_pos)

([8, 12, 14, 15, 19, 22], 90)

Про ранжирование при отрицательных значениях я смотрела тут:
- http://statistica.ru/local-portals/medicine/w-kriteriy-uilkoksona/
- https://medstatistic.ru/methods/methods3.html

Но там данные для попарных разностей для сравнения двух связных выборок, поэтому я не стала  анализировать отдельно отрицательные и положительные показатели, только их сумму

In [11]:
#По таблице находим квантили

wilcoxon_11_11_05 = 100.0
wilcoxon_11_11_95 = 2 * 11* (11 + 11 + 1)/2 - 100
wilcoxon_11_11_95

153.0

- Доверительная область: 􏰛(-∞, 153)
- Критическая область: (153; +∞)
- Значение статистики для суммы рангов отрицательных и положительных значений (42 + 90) = 132 находится в доверительной области - принимаем гипотезу Н0 на уровне значимости 0.05

Т.к. n + m > 20, то можно аппроксимировать распределение статистики W гауссовским 
распределением.

Вычислим E и D

In [12]:
E = 11 * (11 + 11 + 1) / 2
D = 11 * 11 * (11 + 11 + 1) / 12

W = (42 + 90 - E) / math.sqrt(D) # для общей суммы рангов

E, D, W

(126.5, 231.91666666666666, 0.36115755925730764)

- Z1−α = Z0,95 = 1,65
- Доверительная область: (-∞; 1.65), критическая область: (1.65; + ∞)
- Значение статистики 0.36 входит в доверительную область, следовательно, принимаем гипотезу Н0 на уровне значимости α = 0.05.

Попробуем применить **критерий Фишера**

- Гипотеза Н0: SX2/ SY2 = 1
- Гипотеза Н1: SY2 / SX2 != 1

In [13]:
X = np.mean(control)
SX2 = 0.1 * sum([(x - X)**2 for x in control])
X, SX2

(1.0545454545454545, 29.76872727272727)

In [14]:
Y = np.mean(experimental)
SY2 = 0.1 * sum([(y - Y)**2 for y in experimental])
Y, SY2

(2.118181818181818, 39.73763636363637)

Статистика Фишера будет иметь вид

In [15]:
F = SY2 / SX2
F

1.3348785791068116

- Т.к. SY2 > SX2, то статистика имеет распределение Фишера F (n − 1; m − 1), а критическая область критерия имеет вид: (f1−α/2(n−1;m−1);+∞), т.е. f0.975(10,10) = **3.72**. Таким образом, критическая область имеет вид (3.72;+∞)
- Статистика находится в доверительной области, следовательно, принимаем гипотезу Н0 на уровне значимости 0.05: дисперсии можем считать одинаковыми


- Теперь предположения, требуемые для применения критерия Стьюдента со статистикой выполнены. Вычислим реализацию статистики **критерия Стьюдента**
- Статистика Стьюдента при справедливости H0 имеет распределение Стьюдента t(m+n−2)=t(20).


- Гипотеза H0: θ = mY − mX = 0
- Гипотеза H1: θ > 0


- По таблице находим 0,95 квантиль распределения Стьюдента t0,95;20 = 2.09.
- Критические области для Т: (2.09; + ∞)
- Доверительная область для Т: (-∞; 2.09)

In [16]:
X, Y

(1.0545454545454545, 2.118181818181818)

Вычислим выборочную дисперсию объединенной выборки

In [17]:
S2 = 1/20 * (
    sum([(x - X)**2 for x in control]) + \
    sum([(y - Y)**2 for y in experimental])
)

S = math.sqrt(S2)
S2, S

(34.753181818181815, 5.895182933394164)

Реализация статистики будет иметь вид

In [18]:
T = (Y - X)/(S * math.sqrt(1/11 + 1/11))
T

0.42313332928660424

- Значение статистики 0.42 входит в доверительную область (-∞; 2.09), следовательно, принимаем гипотезу Н0 на уровне значимости 0.05

In [19]:
# Проверим расчет полученной статистики

from scipy import stats
stats.ttest_ind(experimental,control)

Ttest_indResult(statistic=0.42313332928660424, pvalue=0.6767131506956277)

Попробуем применить **критерий Ансари–Брэдли**

- Пусть показатели изменения скорости выполнения задания в контрольной группеявляются выборкой Xm = \[X1,...,Xm\]T объема m = 11, порожденной непрерывной случайной величиной X, а показатели изменения скорости выполнения задания в экспериментальной группе - выборкой Yn = \[Y1,...,Yn\]T объема n = 11, порожденной непрерывной случайной величиной Y
- Тогда гипотеза H0 : ∆ = 1 будет означать, что выборки Xm и Yn однородны, и применение препарата не повлияло на скорость выполнения задания. В качестве альтернативной гипотезы в этой задаче выберем HA : ∆ > 1, так как справедливость этой альтернативы означает, что скорость выполнения задания изменилась больше в экспериментальной группе по сравнению с контрольной группой.

In [20]:
control_ranges = [x[0][0] + 1 for x in ranges if x[0][1][1] == 'c' ]
control_ranges

[1, 4, 5, 7, 9, 10, 11, 16, 17, 20, 21]

In [21]:
A = sum([
    23/2 - abs(x - 23/2) for x in control_ranges
])
A

65.0

In [22]:
E = (11 * (22 + 2)) / 4
D = 11 * 11 * 24 * 20 / (48 * 21)
E, D

(66.0, 57.61904761904762)

In [23]:
A_approx = (A - E) / math.sqrt(D)
A_approx

-0.13173978860172217

Критическая область (u_1-α; +∞) для уровня значимости α = 0,05 имеет вид (1,65; +∞). Значение статистики входит в доверительную область, следовательно, принимаем гипотезу Н0 на уровне значимости 0.05

- Проверим гипотезу об однородности выборок Xm и Yn при помощи **критерия Колмогорова–Смирнова**. 
- Так как у нас нет априорной информации о типе неоднородности, следует проверить гипотезу H0 вида H0:FX(t)=FY(t), ∀t∈R1 против альтернативной гипотезы HA общего вида H1:∃t∈R1, такое что FX(t)̸=FY(t). 

In [24]:
from scipy.stats import ks_2samp
ks_2samp(control, experimental)

Ks_2sampResult(statistic=0.2727272727272727, pvalue=0.832587690946824)

- Согласно таблицам, K0,95 = 1,36. Критическая область имеет вид (1.36; +∞) 
- Таким образом, значение статистики попадает в доверительную область, следовательно, принимаем гипотезу Н0 об однородности выборок на уровне значимости 0.05